# Genetic architecture of transmission stage production and virulence in schistosome parasites

## Abstract
Both theory and experimental data from multiple pathogens suggest that the production of transmission stages should be strongly associated with virulence, but the genetic bases of parasite transmission/virulence traits are poorly understood. In the blood fluke *Schistosoma mansoni*, parasite genotypes show extensive variation in numbers of cercariae larvae shed from infected snails. Furthermore, high shedding parasites cause high mortality to snails while low shedding parasites cause low mortality, consistent with expected trade-offs between parasite transmission and virulence. To understand the genetic basis of transmission stage production/virulence, we conducted reciprocal crosses between schistosomes from two laboratory populations that differ 8-fold in cercarial shedding and in their virulence to inbred snail hosts. Each parasite generation, we determined four-week cercarial shedding profiles in inbred *Biomphalaria glabrata* snails infected with single parasite larvae. We sequenced the whole genome of the F0 parents and the exome of the F1 progeny and 188 F2 progeny from each cross, and used linkage mapping to reveal quantitative trait loci (QTLs) underlying transmission stage production. Cercarial production is polygenic: we found three major QTLs on chromosome 1, 3 and 5 (Log-of-the-odds (LOD) = 5.61, 8.19, 6.25) and two minor QTLs on chromosome 2 and 4. These QTLs act additively and explained 28.56% of the phenotypic variation in cercarial shedding. Alleles inherited from the high and low shedding parents were co-dominant at all QTLs, except for chr. 1 and chr. 4 where the “high cercarial shedding” allele is recessive. These results demonstrate that the genetic architecture of key traits directly relevant to schistosome ecology can be dissected using classical linkage mapping approaches, and set the stage for fine mapping and functional validation of the genes involved using the growing armory of functional and cell biology tools available for this parasite.

## Code

The code used to analyze the data presented in the mansucript is available in this repository.

## Prerequisites

Two dependencies must be installed before running the Jupyter notebook:
* A [conda](https://docs.conda.io/en/latest/) distribution, like [miniconda](https://docs.conda.io/en/latest/miniconda.html),
* [Jupyter Notebook](https://jupyter.readthedocs.io/en/latest/install.html) with a [bash kernel](https://github.com/takluyver/bash_kernel).

Once this is done, run the first cell of the Jupyter notebook from this repository to install a dedicated conda environment and related R packages.
